File: IP2LitePerl/Perl/AccessLogger.pl
#!C:/Perl64/bin/perl

=for comment

Created on Jul 28, 2018

@author:  Benjamin Dillard
@title:   Lead Full Stack Developer / Solutions Architect
@email:   bdillard@optonline.net
@website: http://innovationcomputing.org    

       Perl Website AccessLogger using IP2Lite DB11
       
       Requires: IP2Lite Database DB11 from: https://lite.ip2location.com
       
       Developed Using: : CGI Apache HTTP Server 2.4.x                                
                          Perl 64
                          MySQL 5.5                                     
=cut

use DBI;
use POSIX;
use Config::Tiny;
use CGI qw(:standard);

# **************************************************************************** #
#                     Open the Perl config file.                               #
# **************************************************************************** #

print "Content-type: text/html\n\n";

$Config = Config::Tiny->read( '..\\Config\\config.ini' );
$Config = Config::Tiny->read( '..\\Config\\config.ini', 'utf8' ); 
$Config = Config::Tiny->read( '..\\Config\\config.ini', 'encoding(iso-8859-1)');

# **************************************************************************** #
#              Get MySQL userid and password from config.ini                   #
# **************************************************************************** #

my $userid   = $Config->{mysql}->{userid};
my $passwrd  = $Config->{mysql}->{password};
my $path     = $Config->{logs}->{path};
my $filename = $Config->{logs}->{filename};

my $logfile  = $path . POSIX::strftime('%Y%m%d', localtime) . $filename;

# ******************************************************************
#                    Get Remote Users Environment Information
# ******************************************************************

sub getEnvironment {
   
   my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst)=localtime(time);
   my $timestamp = sprintf ( "%04d%02d%02d %02d:%02d:%02d",
                             $year+1900,$mon+1,$mday,$hour,$min,$sec);
                             
   return ("IPAddress = "    . $ENV{REMOTE_ADDR}     . " | " . 
           "Access Date = "  . $timestamp            . " | " .
           "User Agent = "   . $ENV{HTTP_USER_AGENT} . " | " .
           "Referer = "      . $ENV{HTTP_REFERER}    . " | " .
           "Query String = " . $ENV{QUERY_STRING});
         
}

# ******************************************************************
#                    Get Remote Users IPAddress Formatted
# ******************************************************************

sub getFormattedIPAddress {

  my $ipaddress = $ENV{REMOTE_ADDR};

  push(@ipaddress, split(/\./, $ipaddress));
  my $ipaddr    = join( "", @ipaddress );
  return ($ipaddr);
}

# **************************************************************************** #
#                          Open MySQL Connection.                              #
# **************************************************************************** #

sub openConnection {
    
    my $dbh;
    $dbh = DBI->connect( 'dbi:mysql:ip2locationdb11', $userid, $passwrd ) ||
    print "Could not connect to database.\n$DBI::errstr\n";
    return($dbh);
}

# **************************************************************************** #
#                  Query MySQL IP2Lite Database.                               #
# **************************************************************************** #

sub getIPAddressLocation {

    my $conn = shift;
    my $ip   = shift;
    
    $sql  =  "Select ip_from  as ip_from,                    " .
    
                              "ip_to        as ip_to,        " .
                              "country_code as country_code, " .
                              "country_name as country_name, " .
                              "region_name  as region_name,  " .
                              "city_name    as city_name,    " .
                              "latitude     as latitude,     " .
                              "longitude    as longitude,    " .
                              "zip_code     as zip_code,     " .
                              "time_zone    as time_zone     " .
             "From ip2locationdb11                           " .
             "Where '" . $ip . "'" . " Between ip_from And ip_to Limit 0, 1";
    
    $sth = $conn->prepare($sql);
    $sth->execute
    or print "SQL Error: $DBI::errstr\n";
    
    while (@row = $sth->fetchrow_array) {
    
      return (" | Country Code = "  . @row[2] .
              " | Country Name = "  . @row[3] .
              " | Region Name = "   . @row[4] .
              " | City Name = "     . @row[5] .
              " | Latitude = "      . @row[6] .
              " | Longitude = "     . @row[7] .
              " | Zip Code = "      . @row[8] .
              " | Time Zone = "     . @row[9]);  
    }
}

# ******************************************************************
#                     Log Access Information
# ******************************************************************

sub writeToLogFile {

  my $env    = shift;
  my $result = shift;

  open(FH, '>>', $logfile) or print "Error opening file" . $!;
 
  print FH $env . $result . "\n";
 
  close(FH);
}

# **************************************************************************** #
#                           Call Sub Routines.                                 #
# **************************************************************************** #

# Get Remote Users Environment 
my $env = &getEnvironment();

# Get Remote Users IPAddress Formatted
my $ip = &getFormattedIPAddress(); 

# Open connection to MySQL Database
my $conn = &openConnection();            

# Get IPAddress Location from IP2Lite
my $result = &getIPAddressLocation($conn, $ip);

# Write Access information to Logfile
&writeToLogFile($env, $result);

# Close MySQL DBI Connection.
$conn->disconnect();

print ("Perl Access Successfully Logged.")

File: IP2LitePython/Python/AccessLogger.py
#!C:/python34/python
'''
Created on Jul 28, 2018

@author:  Benjamin Dillard
@title:   Lead Full Stack Developer / Solutions Architect
@email:   bdillard@optonline.net
@website: http://innovationcomputing.org    

       Python Website AccessLogger using IP2Lite DB11
       
       Requires: IP2Lite Database DB11 from: https://lite.ip2location.com
       
       Developed Using: : CGI Apache HTTP Server 2.4.3                                 
                          Python 34
                          MySQL 5.5                                     
'''
import os
import cgi
import configparser
import mysql.connector
from mysql.connector import Error
from datetime import datetime

class AccessLogger:

    def __init__(self):
        
        print ("Content-type:text/html\r\n\r\n")
        
    # ******************************************************************
    #         Get AccessLogger Configuration Paramaters                *
    # ******************************************************************
        
        Config = configparser.ConfigParser()
        Config.read("..\\Config\\config.ini")
        
        dbhost       = Config.get('mysql', 'host')
        database     = Config.get('mysql', 'database')
        dbuserid     = Config.get('mysql', 'userid')
        dbpassword   = Config.get('mysql', 'password')
        
        logpath      = Config.get('logs', 'path')
        logname      = Config.get('logs', 'filename')
        
        current_time = datetime.now()
        logfile = ("%s_%s_%s" % (current_time.year, current_time.month, current_time.day) + logname)
        logfile = logpath + logfile
        
        # Get Remote Users Environment 
        env = AccessLogger.getEnvironment(self)
        
        # Get Remote Users IPAddress Formatted
        ipAddress = AccessLogger.getFormattedIPAddress(self) 
        
        # Open connection to MySQL Database
        conn = AccessLogger.openConnection(self, dbhost, dbuserid, dbpassword, database)
        
        # Get IPAddress Location from IP2Lite
        result = AccessLogger.getIPAddressLocation(self, conn, ipAddress)
        
        # Write Access information to Logfile
        AccessLogger.writeToLogFile(self, env, result, logfile)
        
        # Close MySQL Database Connection
        AccessLogger.closeConnection(self, conn)
        
        print ("Python Access Successfully Logged.")
        
# ******************************************************************
#                    Get Remote Users Environment Information
# ******************************************************************
        
    def getEnvironment(self):
        
        time = datetime.now()   
        date = ('{}'.format(time))
        
        return ("IPAddress = "    + cgi.escape(os.environ["REMOTE_ADDR"])     + " | " + 
                "Access Date = "  + date                                      + " | " +
                "User Agent = "   + cgi.escape(os.environ["HTTP_USER_AGENT"]) + " | " +
                "Referer = "      + os.path.realpath('.')                     + " | " +
                "Query String = " + cgi.escape(os.environ["QUERY_STRING"]))
         
        
# ******************************************************************
#                    Get Remote Users IPAddress Formatted
# ******************************************************************
        
    def getFormattedIPAddress(self):
        
        ipaddress = cgi.escape(os.environ["REMOTE_ADDR"])
        
        ipaddress = ipaddress.split(".")
        ipaddress = ''.join(ipaddress)
         
        return ipaddress
        
# ******************************************************************
#                   Open MySQL Database Connection
# ******************************************************************
        
    def openConnection(self, host, userid, password, database):
        
        try:
            conn = mysql.connector.connect(host=host,
                                           database=database,
                                           user=userid,
                                           password=password)
            if conn.is_connected():
                return conn
                
        except Error as e:
            print("MySQL Connection Error. Error = " + e)
            
# ******************************************************************
#         Get IP2Lite IPAddress Location Information
# ******************************************************************
        
    def getIPAddressLocation(self, conn, ipAddress):
        
        try:
        
            query = ("Select ip_from      as ip_from,      " +
                            "ip_to        as ip_to,        " +
                            "country_code as country_code, " +
                            "country_name as country_name, " +
                            "region_name  as region_name,  " +
                            "city_name    as city_name,    " +
                            "latitude     as latitude,     " +
                            "longitude    as longitude,    " +
                            "zip_code     as zip_code,     " +
                            "time_zone    as time_zone     " +
                     "From ip2locationdb11                 " +
                     "Where '" + ipAddress + "'" + " Between ip_from And ip_to Limit 0, 1")
            
            cursor = conn.cursor()
            
            cursor.execute(query)
        
            result = cursor.fetchall()
            
            if not cursor.rowcount:
                return " | Website Accessed but NO IP Information found on IP2Lite DB11 Database"
                
            else:
            
                for row in result:
                    
                    return (" | Country Code = "  + (''.join(map(str, [ row[2] ]))) +
                            " | Country Name = "  + (''.join(map(str, [ row[3] ]))) +
                            " | Region Name = "   + (''.join(map(str, [ row[4] ]))) +
                            " | City Name = "     + (''.join(map(str, [ row[5] ]))) +
                            " | Latitude = "      + (''.join(map(str, [ row[6] ]))) +
                            " | Longitude = "     + (''.join(map(str, [ row[7] ]))) +
                            " | Zip Code = "      + (''.join(map(str, [ row[8] ]))) +
                            " | Time Zone = "     + (''.join(map(str, [ row[9] ]))))         
                             
        except Error as e:
            print ("MySQL Error has occurred in AccessLogger - Exception = " + e)
        
        finally:  
            cursor.close()
    
# ******************************************************************
#                     Log Access Information
# ******************************************************************
        
    def writeToLogFile(self, env, result, logfile):
        
        try:
        
            with open(logfile, 'a') as logfile:
            
                logfile.write(env + result)
                logfile.write('\n')
        
        except Error as e:
            print ("Logfile error has occured - " + e)
# ******************************************************************
#                  Close MySQL Database Connection             
# ******************************************************************
      
    def closeConnection(self, conn):
        
        conn.close()
        
# ******************************************************************
#           Instantiate AccessLogger Class via Constructor
# ******************************************************************
            
AccessLogger()

